### ISSUE TITLE
Color-correcting Eddie boss rooms

### ISSUE THREAD

**Issue by Polymega** [elishacloud/Silent-Hill-2-Enhancements#9]
@Bigmanjapan

Hi friend, 
Playing PC and PS2 versions side-by-side the Eddie boss rooms are discolored and the lighting is rather too different. 

## Eddie boss small room

The biggest issue is the first (small) room: Both the colors are completely off and the lighting is much more dim in the PC version: 

### PC
![sh2pc 2018-06-10 13-23-47](https://user-images.githubusercontent.com/27395260/41204359-8ea48ef4-6cb1-11e8-85a3-a962ccfb2fa2.png)

### PS2
![image](https://user-images.githubusercontent.com/27395260/41204358-8b6acf78-6cb1-11e8-9862-919ffd22d04c.png)

I was able to change this room's colors to better match. It's not perfect but much closer. Although, I don't know if these are the right/best values to change.

Description | Address | Original Value | New Value
-- | -- | -- | --
Room coloring (?) R | 00942A60 (float) | 4 | 2
Room coloring (?) G | 00942A64 (float) | 4 | 4
Room coloring (?) B | 00942A68 (float) | 4 | 2

Which gives you this result: 

![sh2pc 2018-06-10 13-28-36](https://user-images.githubusercontent.com/27395260/41204414-3258887a-6cb2-11e8-919a-98ceeaf58d1a.png)

So the colors are starting to look good but now the room is overall too dark. I don't know what addresses to find/change to now increase its brightness? 

## Eddie boss large room

This room, for the most part, is pretty spot-on in terms of coloring. **Using this room's default values (4, 4, 4) keeps the walls and meat the same coloring between the two.** 

### PC 
![sh2pc 2018-06-10 13-36-50](https://user-images.githubusercontent.com/27395260/41204511-5ba7172c-6cb3-11e8-9826-443960e03d3c.png)

### PS2
![image](https://user-images.githubusercontent.com/27395260/41204474-e439ebba-6cb2-11e8-9c4f-26cc23d1c451.png)

The only difference I noticed is that the **center of the floor** isn't green/blue-tinted like the PS2 version. 

### PC 
![sh2pc 2018-06-10 13-34-30](https://user-images.githubusercontent.com/27395260/41204488-0d7766c4-6cb3-11e8-9911-d30cfb1d9f1f.png)

### PS2
![image](https://user-images.githubusercontent.com/27395260/41204494-1b4a16ac-6cb3-11e8-8d2c-bf527620178e.png)

So I was wondering if there's an overhead light in this room whose RGB values can be changed? 

--------------------

Lastly, don't sweat too much as I am **not** going to go over all of the game's environment colorings with a fine-tooth comb. I would like to only point out the _majorly_ different areas such as this one.

**Comment by Bigmanjapan** [comment]
> I don't know what addresses to find/change to now increase its brightness?

This could have been an easy fix if the PC port actually had a functional brightness adjustment. Only now I noticed that "Brightness Level" menu in Options does nothing while it should (as seen on E3 and 20228GH versions in the next video) adjust overall brightness of the game screen.

https://youtu.be/HoZkU6o_A5Q

> Although, I don't know if these are the right/best values to change.

This game address management is a mess. A lot of rooms use different addresses with different type of values and different ranges on those values to control similar things like colour / brightness. Especially the first toilet location which has it's own set of addresses for almost everything (in "SH2 Making of" narrator says that it was the first location that was modeled, maybe that's the reason). I guess you use any addresses that actually work in this particular location. Another issue is that turning the flashlight on / off resets the values back to default state but that's a minor issue.

I give you this CE table with multiple colour / brightness addresses to experiment with for now (try using negative values on float type values): 

[SH2_brightness.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2091091/SH2_brightness.zip)
https://youtu.be/vPRM59uQunw

I will experiment with it myself also in a few days. 

Another point is, maybe it's easier to change brightness / colors on the actual textures in .map files?

**Comment by Polymega** [comment]
> This could have been an easy fix if the PC port actually had a functional brightness adjustment. Only now I noticed that "Brightness Level" menu in Options does nothing while it should (as seen on E3 and 20228GH versions in the next video) adjust overall brightness of the game screen.
> 
> https://youtu.be/HoZkU6o_A5Q

lol wow... this is also silly. I tried messing around with PC brightness myself and noticed... it changes the noise effect settings but not the actual game brightness. ??

PC Brightness Level | 00942CC5 (byte) | 00942CC0 (byte)
-- | -- | -- 
0 | 255 | 91
1 | 255 | 95
2 | 240 | 96
3 | 222 | 96
4 | 210 | 96
5 | 195 | 96
6 | 170 | 96
7 | 130 | 96

This is good to know, though. It tells me the settings for the [PS2-like noise effect](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/8) need to be "locked" as they'll change otherwise if someone changes the brightness levels. 

Also, @elishacloud could probably correctly answer this one but, from what I understand, the PC brightness adjustments actually _dim/brighten your monitor_ and not just the game. Although, [looking at Elisha's notes](https://github.com/elishacloud/Silent-Hill-2-Enhancements#features) on the matter this may only happen when WineD3D is in use but I have a dual screen setup at home and whenever I change the game's brightness my second monitor changes as well. 

> I guess you use any addresses that actually work in this particular location. Another issue is that turning the flashlight on / off resets the values back to default state but that's a minor issue.
> 
> I give you this CE table with multiple colour / brightness addresses to experiment with for now (try using negative values on float type values):

Copy that and thank you. I will begin experimenting myself as well then. 

> Another point is, maybe it's easier to change brightness / colors on the actual textures in .map files?

It won't fix the issue for the Eddie boss room. We could make the texture's pure white but so long as the light sources inside this room are dimmed the polygon's with these white textures will also still be dimmed, if that makes sense.

**Comment by Polymega** [comment]
I wonder how WineD3D is smart enough to know to adjust your monitor's brightness whenever you change the slider in the game? Was there specialty code written within WineD3D _just_ for this game to fix that? Or is SH2 PC using some sort of universal caller that WineD3D picks up on and knows to adjust accordingly? I also wonder, if that's the case, if WineD3D could adjust the brightness _only_ for the game instead of your entire monitor?

**Comment by Polymega** [comment]
@AeroWidescreen this is a continuation from [your post here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/8#issuecomment-396721078): 

> The brightness control does work like the PS2 version for me. Apparently it's a fullscreen only feature, so it won't work in windowed mode, but it breaks after alt-tabbing.

Nice catch! Yeah, I "print screened" a brightened game image and, sure enough, pasting it into Paint shows the game is still at default levels (so your monitor is just being brightened to compensate). This is the same as what WineD3D does except it makes it work in windowed mode, too. 

So... with these things figured out and considering the Eddie boss "small" room is the only area in the game with any majorly noticeable brightness differences maybe we just aim to recolor the rooms but leave the small room's brightness levels as-is? 

Unlike something like the flashlight or sounds which are pivotal assets to the _entire_ game, I'm not sure trying to adjust the levels of this one room is worth the man power when it still looks presentable after just being recolored. I mean, unlike the [lighting transition bugs](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/2), this room's lighting still looks and functions fine; James isn't wigging out or magically turning bright/dark in the room. Like Bigmanjapan said [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/9#issuecomment-396327020), if it was as simple as changing a few values (that should work regardless if the game is windowed or not) it'd be one thing but this looks to be a whole other can of worms. Thoughts Aero? @Bigmanjapan ?

**Comment by Polymega** [comment]
> Another point is, maybe it's easier to change brightness / colors on the actual textures in .map files?

I reached out to Amy who runs [Alchemilla Hospital](http://alchemillahospital.net/) and is heavily involved with [using available tools to front-end mod the game](https://www.youtube.com/user/AlchemillaHospital1/videos) (such texture swapping). There's unfortunately no way to swap textures for the .map files to her knowledge: 

![image](https://user-images.githubusercontent.com/27395260/41485512-d2d8d66a-70ae-11e8-8b00-814762b2b458.png)

**Comment by elishacloud** [comment]
> I wonder how WineD3D is smart enough to know to adjust your monitor's brightness whenever you change the slider in the game? Was there specialty code written within WineD3D just for this game to fix that? Or is SH2 PC using some sort of universal caller that WineD3D picks up on and knows to adjust accordingly?

AFAIK, there is no specific code in WineD3D for this game.  It looks like what is happening is that WineD3D sees some API call from SH2 and then adjusts the whole monitor brightness.  However, there also appears to be a bug in WineD3D because it does not reset the brightness after the game exits.  I put a fix in for this bug.  You can see it [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/commit/1d91935258e07f3e1201721a551c8dc19c467f5d#diff-cbacad7149b17638b87b6e619c82fa6dR108).

BTW: I suspect this happens when calling the [SetGammaRamp](https://docs.microsoft.com/en-us/windows/desktop/api/d3d9/nf-d3d9-idirect3ddevice9-setgammaramp) API, though I have not tested my theory.

> I also wonder, if that's the case, if WineD3D could adjust the brightness only for the game instead of your entire monitor?

WineD3D tries to convert all the API's from DirectX to OpenGL.  For this API they probably change the whole screen because there is no equivalent API in OpenGL (though I don't really know anything about OpenGL).  In fact this is a legacy API that goes all the way back to DirectX 1 (AFAIK).  See [here](https://docs.microsoft.com/en-us/windows/desktop/api/Ddraw/nn-ddraw-idirectdrawgammacontrol).  Newer versions of DirectX don't have this API.  And there are only a few games that use it.  WineD3D could probably adjust this only for the game window but it may not be worth it for them.

See [this page](https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/using-gamma-correction) for more details on GammaControl.

**Comment by elishacloud** [comment]
I guess I should have tested this before responding, but my theory was correct.  The "dark screen" issue that WineD3D causes can be fixed by remarking [this](https://github.com/crosire/d3d8to9/blob/52a32ef4d8f569ec299bcad13f1a531ce12504c1/source/d3d8to9_device.cpp#L251) line in [d3d8to9](https://github.com/crosire/d3d8to9).

**Comment by Bigmanjapan** [comment]
Sorry for being out of the loop for quite some time, had to visit Gulag. 

Very interesting nuaince about brightness level actually working in fullscreen. I got so used to playing SH2 in windowed mode that I forget to check that. Still, it seems that PC brightness is only an overall brightness thing, changing separate RGB channels even in fullscreen doesn't affect colour scheme the way it does on PS2.

@Polymega regarding what Amy said — I'm not sure I understand what you mean by extracting-importing textures from/to .map files, because I certainly can do extracting-importing in sh_texture_explorer_2.1.

![1](https://user-images.githubusercontent.com/31383389/42344835-be60e0d2-80c7-11e8-8daa-7b943823b20d.png)

Here I extracted two wall textures, changed their colour balance and imported them back to .map files. Have you meant something else?

Also, it seems that Xbox versions share this colour change in Eddie's freezer room with PC version: https://youtu.be/orrtK9GdThc?t=8627

**Comment by Polymega** [comment]
> @Polymega regarding what Amy said — I'm not sure I understand what you mean by extracting-importing textures from/to .map files, because I certainly can do extracting-importing in sh_texture_explorer_2.1.

!!! I was not aware the software could extract and re-import textures for .map files! (I thought it only did this for .mdl files.) This changes everything then! Let me fiddle with this over the weekend. 

Can you send me the link to SH Texture Explorer 2.1? Just so I'm 100% sure I'm using the same one you are using? Great job with this! 

Also, to catch you up to speed: The Audio Enhancement Pack has been released as you know; and Aero is helping Elisha implement your rowboat animation fix solution, cemetery quick load fix solution, increased front draw distance solution, and will also help to implement the PS2-styled noise filter.

**Comment by Bigmanjapan** [comment]
Here you go: [sh_texture_explorer_2.1.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2168054/sh_texture_explorer_2.1.zip)

>Also, to catch you up to speed

Very nice to hear. I have more time now and will continue dealing with mannequin / flashlight issues.

**Comment by Polymega** [comment]
Thank you sir. I edited my post right before you responded but can you tell me the .map file names for the rooms in question here?

**Comment by Bigmanjapan** [comment]
First and second freezer rooms' textures are contained in .data\bg\ps\ps189.map and ps193.map.

**Comment by Polymega** [comment]
Well, I feel completely embarrassed right about now. I always knew the SH Texture Explorer could extract/re-import environment textures but for some stupid reason I thought we had to edit the .cam files this whole time and not the .map files. Wow that's super embarrassing. 

Not only that, but I'd like to apologize for [my comment here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/9#issuecomment-396334248) (and down below) Bigmanjapan as, at the time, I had no clue how the lighting actually worked in this game. 

> It won't fix the issue for the Eddie boss room. We could make the texture's pure white but so long as the light sources inside this room are dimmed the polygon's with these white textures will also still be dimmed, if that makes sense.

Thanks to the Cheat Table you provided me I have a better grasp of how the lighting affects things now. 

-------------------------

So... on that note. I have the first (small) room done! This was surprisingly deceptive to work on because the colors you edit in the TGA files look **nothing** like they do once they're in the game. So it was a ton of trial and error. 

Attached below is the **ps189.map** file edited: 
[ps189.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2168908/ps189.zip)

There also needs to be address values edited to achieve the desired effect. **For this room only** the following values need changed: 

Description | Address | Original Value | New Value
-- | -- | -- | --
PC location textures colour R (set 3) | 00942A60 (float) | 4 | 2.5
PC location textures colour G (set 3) | 00942A64 (float) | 4 | 2
PC location textures colour B (set 3) | 00942A68  (float) | 4 | 2

Here are some comparison shots: 

### PlayStation 2
![image](https://user-images.githubusercontent.com/27395260/42358256-acfa27e0-80a9-11e8-90a3-1e1634297c79.png)

### PC Adjusted 
![sh2pc 2018-07-05 23-17-49](https://user-images.githubusercontent.com/27395260/42358260-b2489d9e-80a9-11e8-9189-150315a89490.png)

### PlayStation 2
![image](https://user-images.githubusercontent.com/27395260/42358286-dacf1126-80a9-11e8-9aff-3a67c8e0bde6.png)

### PC Adjusted 
![sh2pc 2018-07-05 23-18-31](https://user-images.githubusercontent.com/27395260/42358290-dddc3678-80a9-11e8-8ac7-018e8f57eb32.png)

-------------------------------

I will work on the second (larger) room soon.

**Comment by Bigmanjapan** [comment]
Nice work. 

Now I wonder if the actual textures in PS2 builds are coloured differently. In some time a PS2 textures extractor/explorer will be made and we will be able to check that.

I guess if @Aero doesn't mind creating a function to change the colour palette in these rooms, then that would be the final solution. Bear in mind that upon pressing Esc game engine will try to return colour palette values to their default state but that should be a minor issue for Aero.

**Comment by Polymega** [comment]
> Now I wonder if the actual textures in PS2 builds are coloured differently. In some time a PS2 textures extractor/explorer will be made and we will be able to check that.

Yeah, the first thing I did was try to extract the PS2 .map file with SH Texture Explorer, haha. Obviously it didn't work but I wanted to try anyway. 

Also Bigmanjapan, do you know the address to disable fog in the **PS2** version? I'd like to see if my floor has been edited correctly but I can't compare because of all the fog on top.

**Comment by Bigmanjapan** [comment]
[PCSX2_SLUS_20228GH_fog.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2170936/PCSX2_SLUS_20228GH_fog.zip)

Set intensity to 0.

**Comment by Bigmanjapan** [comment]
@Polymega 

[PCSX2_SLUS_20228GH_fog.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2170956/PCSX2_SLUS_20228GH_fog.zip)

Set intensity to 0. It works with outdoors fog, haven't tested these addresses in the freezer rooms.
There are also addresses to disable noise grain and shadows just in case.

**Comment by Polymega** [comment]
@Bigmanjapan there's definitely some sort of special light that perhaps is only meant for this room. The texture on the floor is solid white. Any chance of finding where the addresses for this are at? 

![sh2pc 2018-07-06 22-08-11](https://user-images.githubusercontent.com/27395260/42405742-368dcdb2-8169-11e8-9661-dffdaac41874.png)

**Comment by Polymega** [comment]
@Bigmanjapan scratch that request. It's crazy how adjusting the RGB levels for other addresses changes the coloring of textures so oddly. I think I found a solution without having to find this special blue light.

**Comment by Bigmanjapan** [comment]
I see. This confusion might be due to a memory shortage on PS2 that developers had to use a limited number of addresses to control colour palettes, therefore assigning different effects for different rooms on the same addresses.

**Comment by Polymega** [comment]
Sorry, I should have clarified: The screenshot above is from the PC version, in case you were thinking that was from the PS2 build. 

Here's what I was able to do with the blue light active: 

### PC Adjusted

![sh2pc 2018-07-07 00-51-45](https://user-images.githubusercontent.com/27395260/42406936-f60bd6f0-817f-11e8-82dc-6e6c6e2117a0.png)

### PlayStation 2

![image](https://user-images.githubusercontent.com/27395260/42406899-4caa1fea-817f-11e8-8c9a-54ce7d789a24.png)

The middle of the floor (right around Eddie) looks good/close now. But near the edges (towards the walls) the PC adjusted version is more saturated vs. the PS2 version. This is because of that blue light influencing color all along the wall's edges. 

I've actually spent the past several hours on this one floor but have to concede that I can't match it any better due to that overcasting blue light. What do you think though?

**Comment by Bigmanjapan** [comment]
> Sorry, I should have clarified: The screenshot above is from the PC version, in case you were thinking that was from the PS2 build.

Sure. I meant that memory management logic was transferred to PC from PS2 with no changes. Therefore PC builds also have this different effects swapping on the same addresses despite PC computers having more available RAM.

> I've actually spent the past several hours on this one floor but have to concede that I can't match it any better due to that overcasting blue light. What do you think though?

I'll try to find a way to disable or adjust it.

**Comment by Polymega** [comment]
> Sure. I meant that memory management logic was transferred to PC from PS2 with no changes. Therefore PC builds also have this different effects swapping on the same addresses despite PC computers having more available RAM.

Ah, okay, that makes sense then. That's also a smart theory.

> I'll try to find a way to disable or adjust it.

Thank you.

**Comment by Bigmanjapan** [comment]
Have you tied setting "PC floor + furniture brightness?" address set to a value of 1? It turns blue light to white, although making some wall elements a bit brighter. 

![default](https://user-images.githubusercontent.com/31383389/42407250-8ffa77aa-81e2-11e8-8f5f-811734c39f22.png)

**Comment by Bigmanjapan** [comment]
Try changing "PC floor + furniture brightness?" address set to a value of 1 or -1, -2, -3 etc and see how it affects the new floor texture, don't use white placeholder texture.

**Comment by Bigmanjapan** [comment]
Have you tried changing "PC floor + furniture brightness?" address set to a value of 1 or -1, -2, -3 etc? If now, see how it affects the new floor texture, don't use white placeholder texture.

**Comment by Polymega** [comment]
> Have you tried changing "PC floor + furniture brightness?" address set to a value of 1-5 or -1, -2, -3 etc? If not, see how it affects the new floor texture, don't use white placeholder texture. 

Thanks. As mentioned prior: I'm getting a better grasp at how the lighting works but I'm still not fully there. After more adjustments I have an even closer approximation. And at this point I'm satisfied with the floor: 

### PC Adjusted

![sh2pc 2018-07-07 11-31-21](https://user-images.githubusercontent.com/27395260/42412387-7db35e26-81d9-11e8-8f9a-177370801a10.png)

### PlayStation 2

![image](https://user-images.githubusercontent.com/27395260/42412386-78d9183c-81d9-11e8-8de5-bd1a2b8dd581.png)

I'm going to go over the other assets in this room next and once I'm done I'll post my adjustments/notes accordingly. Thanks again, Bigmanjapan!

**Comment by Polymega** [comment]
@AeroWidescreen @elishacloud 

Editing for the Eddie boss rooms is complete. The modifications will involve both replacing .map files ([in the same spirit as what the WSF will do](https://github.com/ThirteenAG/WidescreenFixesPack/tree/master/data/SilentHill2.WidescreenFix/scripts/bg)) and adjusting address values while in these particular rooms. 

As Bigmanjapan pointed out it's worth noting that pressing "Esc" while in these rooms the game will try to revert the address changes, so something to be mindful of. 

[The fixes/notes can be found in this post here.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/9#issuecomment-402914130)  

Thank you

**Comment by AeroWidescreen** [comment]
Good work guys, I'll get to it as soon as I can.

**Comment by AeroWidescreen** [comment]
@elishacloud

```
00658674 scripts/bg/ps189.map // "small room" file path; max length is 20 Characters
0065842C scripts/bg/ps193.map // "large room" file path; max length is 20 Characters
```

I recommend using these strings for a few reasons. First, it's exactly 20 characters (same length as original) so nothing extra is required to get that working. Second, the widescreen fix will also be using custom files that will be stored in "scripts/bg/". This keeps things organized and consistent. The third reason is that no files will need to be replaced using this method.

---


For the RGB, I found what appears to be exclusive memory addresses that contain the float values for these two rooms. The game loads these values and moves them over to the addresses used in real-time. I did some quick testing and didn't find any issues, however, **a full playthrough is neccesary** to ensure these values aren't being used anywhere else or being overwritten.

I have provided a cheat table for troubleshooting: [Download](https://mega.nz/#!UCpWTSTC!YunBErxIeLXmVOVy4imfvq9ZAPfIaLqkY_dA67Gomxc)

Small Room (ps189)

```
007FBBD0 40200000 // "Small Room" Location Texture Red; 2.5 float
007FBBD4 40000000 // "Small Room" Location Texture Green; 2.0 float
007FBBD8 40000000 // "Small Room" Location Texture Blue; 2.0 float
```

Large Room (ps193)

```
007FBDC0 4049999A // "Large Room" Location Texture Red; 3.15 float
007FBDC4 40333333 // "Large Room" Location Texture Green; 2.8 float
007FBDC8 40333333 // "Large Room" Location Texture Blue; 2.8 float

007FBC00 00000000 // "Large Room" Floor Red; 0.0 float
007FBC04 BDCCCCCD // "Large Room" Floor Green; -0.1 float
007FBC08 BDCCCCCD // "Large Room" Floor Blue; -0.1 float
```

**Comment by elishacloud** [comment]
This issue was implemented with the check-in above and I added some fixes to it with [a9da7c4](https://github.com/elishacloud/Silent-Hill-2-Enhancements/commit/a9da7c49702d21729757da07942ba3251b28cfa5) and [cade5ca](https://github.com/elishacloud/Silent-Hill-2-Enhancements/commit/cade5ca011a6f84de836cabb43819b00cc64bdc4).

Closing this issue!
